### ISSUE TITLE
Release 9.8.0

### ISSUE THREAD

**Issue by jeddy3** [stylelint/stylelint#3779]
- [ ] stylelint release
- [ ] stylelint-demo update
- [ ] stylelint.io update
- [ ] tweet

I should have time to put out a release this evening.

It'd be good to get reviews on https://github.com/stylelint/stylelint/pull/3748 and https://github.com/stylelint/stylelint/pull/3775 before I do.

cc / @stylelint/core

**Comment by jeddy3** [comment]
Even though `master` looks good on Travis, I'm getting failing tests locally:

```
Summary of all failing tests
 FAIL  lib/__tests__/standalone-cache.test.js
  ● standalone cache › cache file is created at $CWD/.stylelintcache

    expect(received).toBe(expected) // Object.is equality

    Expected: true
    Received: false

      66 |       .then(cacheFile => {
      67 |         // Ensure cache file contains only linted css file
    > 68 |         expect(typeof cacheFile[validFile] === "object").toBe(true);
         |                                                          ^
      69 |         expect(typeof cacheFile[newFileDest] === "undefined").toBe(true);
      70 |       });
      71 |   });

      at toBe (lib/__tests__/standalone-cache.test.js:68:58)
      at tryCatch (node_modules/es6-promise/dist/lib/es6-promise/-internal.js:195:12)
      at invokeCallback (node_modules/es6-promise/dist/lib/es6-promise/-internal.js:210:13)
      at publish (node_modules/es6-promise/dist/lib/es6-promise/-internal.js:178:7)
      at flush (node_modules/es6-promise/dist/lib/es6-promise/asap.js:94:5)

  ● standalone cache › only changed files are linted

    expect(received).toBe(expected) // Object.is equality

    Expected: true
    Received: false

       97 |       })
       98 |       .then(cachedFiles => {
    >  99 |         expect(typeof cachedFiles[validFile] === "object").toBe(true);
          |                                                            ^
      100 |         expect(typeof cachedFiles[newFileDest] === "object").toBe(true);
      101 |       });
      102 |   });

      at toBe (lib/__tests__/standalone-cache.test.js:99:60)

  ● standalone cache › invalid files are not cached

    expect(received).toBe(expected) // Object.is equality

    Expected: true
    Received: false

      152 |       })
      153 |       .then(cachedFiles => {
    > 154 |         expect(typeof cachedFiles[validFile] === "object").toBe(true);
          |                                                            ^
      155 |         expect(typeof cachedFiles[newFileDest] === "undefined").toBe(true);
      156 |       });
      157 |   });

      at toBe (lib/__tests__/standalone-cache.test.js:154:60)

  ● standalone cache › files with syntax errors are not cached

    expect(received).toBe(expected) // Object.is equality

    Expected: true
    Received: false

      168 |         const cachedFiles = JSON.parse(fileContents);
      169 |
    > 170 |         expect(typeof cachedFiles[validFile] === "object").toBe(true);
          |                                                            ^
      171 |         expect(typeof cachedFiles[newFileDest] === "undefined").toBe(true);
      172 |       });
      173 |   });

      at toBe (lib/__tests__/standalone-cache.test.js:170:60)

  ● standalone cache uses cacheLocation › cacheLocation is a file

    expect(received).toBe(expected) // Object.is equality

    Expected: true
    Received: false

      261 |       })
      262 |       .then(cacheFile => {
    > 263 |         expect(typeof cacheFile[validFile] === "object").toBe(true);
          |                                                          ^
      264 |       });
      265 |   });
      266 |   it("cacheLocation is a directory", () => {

      at toBe (lib/__tests__/standalone-cache.test.js:263:58)

  ● standalone cache uses cacheLocation › cacheLocation is a directory

    expect(received).toBe(expected) // Object.is equality

    Expected: true
    Received: false

      284 |       })
      285 |       .then(cacheFile => {
    > 286 |         expect(typeof cacheFile[validFile] === "object").toBe(true);
          |                                                          ^
      287 |       });
      288 |   });
      289 | });

      at toBe (lib/__tests__/standalone-cache.test.js:286:58)


Test Suites: 1 failed, 291 passed, 292 total
Tests:       6 failed, 14222 passed, 14228 total
Snapshots:   6 passed, 6 total
Time:        61.198s
Ran all test suites.
npm ERR! Test failed.  See above for more details.
```

It looks like node@10 on appveyor is failing the same tests too: https://ci.appveyor.com/project/stylelint/stylelint/branch/master/job/759779kf1trxfg5l

@stylelint/core Does anyone have any idea where this regression could be coming from?

**Comment by hudochenkov** [comment]
It's [`flat-cache@1.3.1`](https://github.com/royriojas/flat-cache), which is a dependency of `file-entry-cache`. After messing with package-lock.json, I was able to force `flat-cache@1.3.0` to be installed version and problem was gone. Perhaps their new dependency [flatted](https://github.com/WebReflection/flatted) giving other results to the same input, thats why our assertions fails.

**Comment by jeddy3** [comment]
Good sleuthing! Does anyone have time to raise this concern upstream in `flat-cache`?

**Comment by evilebottnawi** [comment]
Hm, we in 2k18 and we don't have lock file :smile: , maybe it is time to fix it?

**Comment by hudochenkov** [comment]
After hours of research, I realized that our cache tests are doing a horrible thing. They based on implementation details of dependency of a dependency of a dependency. That's why they broke.

`flat-cache@1.3.1` is a dependency of `file-entry-cache@2.0.0`, which we use to write and read `.stylelintcache`. In v1.3.1 flat-cache replaced their dependency which writes and reads JSON. That sub-dependency put info in `.stylelintcache`. The format of that data is different. Explanation of the new format: https://github.com/WebReflection/flatted/blob/master/SPECS.md.

`.stylelintcache` created by `flat-cache@1.3.1`:

```json
[{"/Users/aleks/projects/stylelint/lib/__tests__/fixtures/cache/valid.css":"1"},{"size":47,"mtime":1498064634000,"hashOfConfig":"2"},"1ai2bgr"]
```

`.stylelintcache` created by `flat-cache@1.3.0`:

```json
{"/Users/aleks/projects/stylelint/lib/__tests__/fixtures/cache/valid.css":{"size":47,"mtime":1498064634000,"hashOfConfig":"1ai2bgr"}}
```

Our tests check this format. While newly created `.stylelintcache` will work, I don't know what happens when users, who have `.stylelintcache` created by the previous version, will run a new version of stylelint. There might be errors due to format differences.

Next steps to tackle the problem:

1. Verify if users, who have `.stylelintcache` in the old format, get errors. I will do this.
2. Fix tests. I won't have time to do this.
3. Depends on findings in point 1, there might be other changes required. I won't have time to do this.

**Comment by evilebottnawi** [comment]
My 50 cents: changes something in `stylelint.config.js` doesn't invalid cache.

**Comment by jeddy3** [comment]
> After hours of research, I realized that our cache tests are doing a horrible thing.

Thank you for doing this!

> Verify if users, who have .stylelintcache in the old format, get errors. I will do this.

Fingers crossed that the cache is just invalidated.

**Comment by hudochenkov** [comment]
> Verify if users, who have `.stylelintcache` in the old format, get errors. I will do this.

Tested. No errors are thrown. The cache file just gets overridden.

> My 50 cents: changes something in `stylelint.config.js` doesn't invalid cache.

Open a new issue, please :)

**Comment by hudochenkov** [comment]
So, we just need to fix tests.

**Comment by jeddy3** [comment]
> So, we just need to fix tests.

I'm next available on Friday evening. I can do a release then if someone finds time to patch the tests before then.

**Comment by ntwb** [comment]
Updating `master` locally and all tests are passing for me without any modifications.

Also a great opportunity to try the _type ahead_ update with `cache` 😏

![screen shot 2018-11-08 at 11 00 57 am](https://user-images.githubusercontent.com/1016458/48169030-b09b0c80-e345-11e8-9948-75e3aeccd843.png)

**Comment by sswebcoder** [comment]
> Updating `master` locally and all tests are passing for me without any modifications.
> 
> Also a great opportunity to try the _type ahead_ update with `cache` 
> 
> ![screen shot 2018-11-08 at 11 00 57 am](https://user-images.githubusercontent.com/1016458/48169030-b09b0c80-e345-11e8-9948-75e3aeccd843.png)

Do you run tests with .stylelintrc or without?

**Comment by hudochenkov** [comment]
@ntwd most likely you didn't update dependencies completely. I bet `npm list flat-cache` gives `1.3.0` :)

**Comment by jeddy3** [comment]
Either way, as @hudochenkov said, we seem to be relying on implementation details of dependency of a dependency of a dependency in our tests. Hopefully, there's a way to tweak the tests so they perform the same purpose yet avoid relying on these implementation details.

**Comment by hudochenkov** [comment]
Surprise! Author of `flat-cache` reverted dependency change 10 hours ago https://github.com/royriojas/flat-cache/blob/master/changelog.md Now it uses an old format again, and out tests pass.

So the release is unblocked. I'll create a technical debt issue.

**Comment by ntwb** [comment]
Yep, it does 🤣

My bad...

```
├─┬ file-entry-cache@2.0.0
│ ├─┬ flat-cache@1.3.0
```

So I deleted `node_modules` and `npm i`'d again, and the tests all passed...

Running `npm list flat-cache` now results in:

```
├─┬ file-entry-cache@2.0.0
│ ├─┬ flat-cache@1.3.2
```

Ha, @hudochenkov, yep, updated `flat-cache` works for me 😏

**Comment by hudochenkov** [comment]
ESLint also have the same problem: https://github.com/eslint/eslint/issues/11062

I've created https://github.com/stylelint/stylelint/issues/3786, so we can move this discussion there.

And again, we can release 9.8.0 now :)

**Comment by jeddy3** [comment]
> And again, we can release 9.8.0 now :)

Cool beans. I can pick this up on Friday evening unless someone beats me to it.
